src/compiler/sys.ts

```diff
@@ -297,6 +304,131 @@ namespace ts {
+
+                    // Node 4.0 `fs.watch` function supports the "recursive" option on both OSX and Windows
+                    // (ref: https://github.com/nodejs/node/pull/2649 and https://github.com/Microsoft/TypeScript/issues/4643)
+                    if (isNode4OrLater() && recursive === true) {
+                        if (recursiveDirWatchers.contains(dirPath)) {
+                            const watcher = recursiveDirWatchers.get(dirPath);
+                            watcher.referenceCount += 1;
+                            return { watcher, isRecursive: true };
+                        }
+                        watchers = recursiveDirWatchers;
+                        options.recursive = true;
+                    }
+                    else {
+                        if (dirWatchers.contains(dirPath)) {
+                            const watcher = dirWatchers.get(dirPath);
+                            watcher.referenceCount += 1;
+                            return { watcher, isRecursive: false };
+                        }
+                        watchers = dirWatchers;
+                    }
+
+                    const watcher: DirWatcher = _fs.watch(dirPath, options, (eventName: string, relativeFileName: string) => fileEventHandler(eventName, relativeFileName, dirPath));
+                    watcher.referenceCount = 1;
+                    watchers.set(dirPath, watcher);
+                    return { watcher, isRecursive: false };
+                }
+
+                function findDirWatcherForFile(filePath: Path): { watcher: DirWatcher, watcherPath: Path, isRecursive: boolean } {
+                    let watcher: DirWatcher;
+                    let watcherPath: Path;
+                    let isRecursive = false;
+                    recursiveDirWatchers.forEachValue(dirPath => {
+                        if (filePath.indexOf(dirPath) === 0) {
+                            watcherPath = dirPath;
+                            watcher = recursiveDirWatchers.get(dirPath);
+                            isRecursive = true;
+                            return;
+                        }
+                    });
+                    if (!watcher) {
+                        const parentDirPath = getDirectoryPath(filePath);
+                        if (dirWatchers.contains(parentDirPath)) {
+                            watcherPath = parentDirPath;
+                            watcher = dirWatchers.get(parentDirPath);
+                        }
+                    }
+                    return { watcher, watcherPath, isRecursive };
+                }
+
+                function addFile(fileName: string, callback: FileWatcherCallback): WatchedFile {
+                    const filePath = toPath(fileName, currentDirectory, getCanonicalPath);
+                    const { watcher } = findDirWatcherForFile(filePath);
+                    if (!watcher) {
+                        addDirWatcher(getDirectoryPath(filePath));

```

Comment:
The returned `watcher` instance is not used. Would ignoring the returned results yield any side effects?